File: _recipes/node/logger.ts
// Recipe: JSON-line structured logger
// Stdlib-only (node:stream, node:process).
// Usage:
//   import { createLogger } from './logger.js';
//   const log = createLogger({ name: 'my-service' });
//   log.info('started', { port: 8080 });
//
// Every call emits one JSON object per line. Custom fields are merged at the
// top level. Error values are serialized as { name, message, stack }.

import { Writable } from 'node:stream';

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

const LEVELS: Record<LogLevel, number> = { debug: 10, info: 20, warn: 30, error: 40 };

export interface LoggerOptions {
  name: string;
  level?: LogLevel;
  stream?: Writable;
}

export interface Logger {
  debug: (message: string, fields?: Record<string, unknown>) => void;
  info: (message: string, fields?: Record<string, unknown>) => void;
  warn: (message: string, fields?: Record<string, unknown>) => void;
  error: (message: string, fields?: Record<string, unknown>) => void;
}

function serializeValue(v: unknown): unknown {
  if (v instanceof Error) {
    return { name: v.name, message: v.message, stack: v.stack };
  }
  return v;
}

export function createLogger({
  name,
  level = 'info',
  stream = process.stderr,
}: LoggerOptions): Logger {
  const threshold = LEVELS[level];

  const emit = (lvl: LogLevel, message: string, fields?: Record<string, unknown>): void => {
    if (LEVELS[lvl] < threshold) return;
    const payload: Record<string, unknown> = {
      timestamp: new Date().toISOString(),
      level: lvl,
      logger: name,
      message,
    };
    if (fields) {
      for (const [k, v] of Object.entries(fields)) {
        payload[k] = serializeValue(v);
      }
    }
    stream.write(`${JSON.stringify(payload)}\n`);
  };

  return {
    debug: (m, f) => emit('debug', m, f),
    info: (m, f) => emit('info', m, f),
    warn: (m, f) => emit('warn', m, f),
    error: (m, f) => emit('error', m, f),
  };
}

File: _recipes/python/logger.py
# Recipe: JSON-line structured logger built on stdlib `logging`
# Stdlib-only.
# Usage:
#   from logger import get_logger
#   log = get_logger("my-service")
#   log.info("started", extra={"port": 8080})
#
# Every log record is emitted as a single JSON object per line containing
# at minimum: timestamp, level, logger, message. Anything you pass via
# `extra={...}` is merged into the JSON object at the top level.
from __future__ import annotations

import json
import logging
import sys
from datetime import datetime, timezone
from typing import IO, Any

_RESERVED = {
    "args", "asctime", "created", "exc_info", "exc_text", "filename",
    "funcName", "levelname", "levelno", "lineno", "message", "module",
    "msecs", "msg", "name", "pathname", "process", "processName",
    "relativeCreated", "stack_info", "thread", "threadName", "taskName",
}


class _JsonFormatter(logging.Formatter):
    def format(self, record: logging.LogRecord) -> str:
        payload: dict[str, Any] = {
            "timestamp": datetime.fromtimestamp(record.created, tz=timezone.utc).isoformat(),
            "level": record.levelname,
            "logger": record.name,
            "message": record.getMessage(),
        }
        for key, value in record.__dict__.items():
            if key in _RESERVED or key.startswith("_"):
                continue
            payload[key] = value
        if record.exc_info:
            payload["exception"] = self.formatException(record.exc_info)
        return json.dumps(payload, default=str)


def get_logger(
    name: str,
    *,
    level: int = logging.INFO,
    stream: IO[str] | None = None,
) -> logging.Logger:
    """Return a logger configured with a single JSON-line StreamHandler.

    Idempotent for a given `name` — repeated calls return the same logger and do
    not duplicate handlers. `stream` defaults to sys.stderr.
    """
    log = logging.getLogger(name)
    log.setLevel(level)
    log.propagate = False
    if not log.handlers:
        handler = logging.StreamHandler(stream or sys.stderr)
        handler.setFormatter(_JsonFormatter())
        log.addHandler(handler)
    else:
        for h in log.handlers:
            h.setLevel(level)
    return log
